lib/Pool.js

```diff
@@ -79,29 +80,52 @@ class Pool {
       // the front of the waitingConnectionWork queue, then either call the
       // callback function provided by the user, or resolve the Promise that was
       // returned to the user.
-      if (this.waitingConnectionWork.length > 0)
-      {
-        let connectionWork = this.waitingConnectionWork.shift();
-
-        // If the user passed a callback function, then call the function with
-        // no error in the first parameter and the connection in the second
-        // parameter
-        if (typeof connectionWork == 'function')
-        {
-          return connectionWork(null, connection);
-        // If the user didn't pass a callback function, we returned a promise to
-        // them. Resolve that promise with the connection that was just created.
-        } else {
-          // Promise (stored resolve function)
-          return connectionWork.resolveFunction(connection);
-        }
+      let connectionWork = this.waitingConnectionWork.shift();
 
       // A connection finished connecting, and there was no work waiting for
       // that connection to be made. Simply add it to the array of
       // freeConnections, and the next time work comes in it can simply be
       // retrieved from there.
-      } else {
+      if (typeof connectionWork == 'undefined')
+      {
         this.freeConnections.push(connection);
+        return;
+      }
+
+      // If the user passed a callback function, then call the function with
+      // no error in the first parameter and the connection in the second
+      // parameter
+      if (typeof connectionWork == 'function')
+      {
+        return connectionWork(null, connection);
+      } 
+
+      // If the user didn't pass a callback function, we returned a promise to
+      // them. Resolve that promise with the connection that was just created.
+      // Promise (stored resolve function)
+      return connectionWork.resolveFunction(connection);
+    });
+
+    // Fires when a connection has errored. If there is no work waiting, then
+    // the pool will simply be empty until the next time work is requested, at
+    // which point there WILL be waiting work and they will get an error.
+    this.connectionEmitter.on('connectionError', (error) => {
+      let connectionWork = this.waitingConnectionWork.shift();
+      if (typeof connectionWork == 'undefined')
+      {
+        return;
+      }
+      // If the user passed a callback function, then call the function with
+      // no error in the first parameter and the connection in the second
+      // parameter
```

Comment:
This comment seems like it was meant for a successful case.

Reply:
Ctrl + C, Ctrl + V strikes again